# Exercise 2.A Fundamentals of pandas

In [5]:
import numpy as np
import pandas as pd

## Series

In [ ]:
aa = [1, 2, 3, 4]
ab = (5, 6, 7, 8)
ac = {9, 10, 11, 12}
ad = {
    'first'   : 13, 
    'second'  : 14, 
    'third'   : 15, 
    'fourth'  : 16
}

In [ ]:
ser1 = pd.Series(aa)
ser2 = pd.Series(ab)
ser3 = pd.Series(list(ac)) #Casts ac as a list because creating a series with a set doesn't work, as a series uses the index to order and identify each value, and sets aren't ordered. 
ser4 = pd.Series(ad)

print(ser1)
print(ser1)
print(ser3)
print(ser4)

0    1
1    2
2    3
3    4
dtype: int64
0    1
1    2
2    3
3    4
dtype: int64
0     9
1    10
2    11
3    12
dtype: int64
first     13
second    14
third     15
fourth    16
dtype: int64


In [70]:
ser5 = 17, 18, 19, 20
#No errors above, I try to make ser5 a list below to create an error
ad = 17, 18, 19, 20
ser5 = pd.Series(ad)
print(ser5)
#No issues assigning, I'm assuming this is part of the workbook errors and they wanted me to see the error trying to create a series with a set. See explanation above

0    17
1    18
2    19
3    20
dtype: int64


## DataFrames from Series

In [71]:
df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


In [72]:
#insert(index column placement, name, series to add)
df1.insert(1, 'ser2', ser2)
df1.insert(2, 'ser3', ser3)

In [73]:
#called here as running above code returns an error as you can't insert an existing column again.
df1


,0,ser2,ser3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [74]:
df1.insert(2, 'ser4', ser4)

In [75]:
df1
#ser4 returns as NaN for all rows, which is probably due to ser4 containing strings rather than integers.

,0,ser2,ser4,ser3
0,1,5,NaN,9
1,2,6,NaN,10
2,3,7,NaN,11
3,4,8,NaN,12


In [76]:
df1.loc[4] = [5, 9, 12, 0]

In [77]:
df1

,0,ser2,ser4,ser3
0,1,5,NaN,9
1,2,6,NaN,10
2,3,7,NaN,11
3,4,8,NaN,12
4,5,9,12.0,0


## Reading Data into a DataFrame

In [78]:
df2 = pd.read_csv('sakila_example.csv')
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [79]:
df2.head() #display top 5 rows by default

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [80]:
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [81]:
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   title        262 non-null    str  
 1   description  262 non-null    str  
 2   category     262 non-null    str  
 3   language     262 non-null    str  
 4   actor_name   262 non-null    str  
dtypes: str(5)
memory usage: 10.4 KB


Some information I can determine from using the info method includes:
- Datatypes used, only str here
- Number of entries, 262
- Number of non-null values in the data, 262 or all values
- Column information:
    - no. of columns
    - column categories
- How much memory the data takes, 10.4 KB

In [82]:
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


Some information I can determine from using the describe method includes:
- Info of top row, can look for a specific row or values.
- !Number of unique values for each column!
- Column names
- Presented closer to what the table appears visually

## Coding a DataFrame

In [83]:
test_data = {
'ID'    :   [1, 2, 3, 4, 5, 6, 7, 8],
'Name'  :   ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
'Age'   :   [25, np.nan, 22, 19, 31, 35, np.nan, 18],
'City'  :   ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
'Score' :   [85, 92, -1, 65, 78, np.nan, 55, 90]
}

In [86]:
df4 = pd.DataFrame(test_data)
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [87]:
df4.to_csv('test_data.csv', index=False)